README.md
# webmanifest-cli

> A generator of webmanifest and icons

[![](https://github.com/ymkz/webmanifest-cli/workflows/check-pull-request/badge.svg)](https://github.com/ymkz/webmanifest-cli/actions?query=workflow%3A"check-pull-request")
[![](https://github.com/ymkz/webmanifest-cli/workflows/healthcheck-master/badge.svg)](https://github.com/ymkz/webmanifest-cli/actions?query=workflow%3A"healthcheck-master")

## Prerequisites

- node >=12

## Install

```sh
npm i -D webmanifest-cli
```

## Usage

```sh
$ webmanifest
$ webmanifest --icon public/icon.svg --manifest public/manifest.webmanifest --outputIcon public/assets --outputManifest public/manifest.webmanifest --sizes 512,384,192,180,152,144,128,96,72
```

---

👤 **ymkz**

- Twitter: [@ymkzly](https://twitter.com/ymkzly)
- Github: [@ymkz](https://github.com/ymkz)


CHANGELOG.md
## [0.1.7](https://github.com/ymkz/webmanifest-cli/compare/v0.1.6...v0.1.7) (2020-05-25)

### Bug Fixes

- **deps:** update dependency meow to v7 ([9455112](https://github.com/ymkz/webmanifest-cli/commit/945511252a1192180e188b707d02cdcddda667db))

## [0.1.6](https://github.com/ymkz/webmanifest-cli/compare/v0.1.5...v0.1.6) (2020-05-18)

### Bug Fixes

- **deps:** update dependency sharp to ^0.25.3 ([19cd3a5](https://github.com/ymkz/webmanifest-cli/commit/19cd3a5ebb0b83f09790ad57a2127621661fc25e))

## [0.1.5](https://github.com/ymkz/webmanifest-cli/compare/v0.1.4...v0.1.5) (2020-05-03)

### Bug Fixes

- **deps:** update dependency meow to ^6.1.1 ([45fbb6d](https://github.com/ymkz/webmanifest-cli/commit/45fbb6dd3354797641b4039c11575486a67fc3de))

## [0.1.4](https://github.com/ymkz/webmanifest-cli/compare/v0.1.3...v0.1.4) (2020-03-27)

### Bug Fixes

- **deps:** update dependency fs-extra to v9 ([2819a16](https://github.com/ymkz/webmanifest-cli/commit/2819a161dca152d6d422e3020cc0c4458ac41620))

## [0.1.3](https://github.com/ymkz/webmanifest-cli/compare/v0.1.2...v0.1.3) (2020-03-21)

### Bug Fixes

- **deps:** update dependency sharp to ^0.25.2 ([3369668](https://github.com/ymkz/webmanifest-cli/commit/3369668f6d388b917d61ced659215127ff3feb17))

## [0.1.2](https://github.com/ymkz/webmanifest-cli/compare/v0.1.1...v0.1.2) (2020-03-20)

### Bug Fixes

- **deps:** update dependency meow to ^6.1.0 ([4fd29ab](https://github.com/ymkz/webmanifest-cli/commit/4fd29abfdd36ac347769aaef594391a0bee1936b))

## [0.1.1](https://github.com/ymkz/webmanifest-cli/compare/v0.1.0...v0.1.1) (2020-03-09)

### Bug Fixes

- **deps:** update dependency sharp to ^0.25.1 ([f112120](https://github.com/ymkz/webmanifest-cli/commit/f1121205d421f10e18f9e93ec4939300810fa6ed))

# [0.1.0](https://github.com/ymkz/webmanifest-cli/compare/v0.0.1...v0.1.0) (2020-03-01)

### Features

- **breaking-change:** upgrade codebase and dependencies ([52b2247](https://github.com/ymkz/webmanifest-cli/commit/52b22478da98f27ae25a31e688cec13035d11f81))


scripts/test.sh
#!/usr/bin/env sh

set -eu

EXAMPLE_ICON="public/icon.svg"
EXAMPLE_WORKSPACE="./example"
PKG_NAME=`node -p "require('./package.json').name"`
PKG_VERSION=`node -p "require('./package.json').version"`
PKG_PACKED=$PKG_NAME-$PKG_VERSION.tgz

if [ -e $PKG_PACKED ]; then
  echo "** Already exists packed module: $PKG_PACKED"
  echo "** Remove old packed file..."
  rm $PKG_PACKED
  echo "** Done, package remove"
fi

echo "+ Building module..."
npm run build

echo "+ Packing module..."
npm pack

if [ -e $EXAMPLE_WORKSPACE ]; then
  echo "** Already exists example workspace: $EXAMPLE_WORKSPACE"
  echo "** Remove old workspace..."
  rm -rf $EXAMPLE_WORKSPACE
fi

echo "+ generate workspace directory"
mkdir -p $EXAMPLE_WORKSPACE/public

echo "+ move test-target module to workspace"
mv $PKG_PACKED $EXAMPLE_WORKSPACE

echo "+ copy necessary icon and manifest"
cp public/icon.svg $EXAMPLE_WORKSPACE/public
cp public/manifest.webmanifest $EXAMPLE_WORKSPACE/public

echo "+ change directory to workspace"
cd $EXAMPLE_WORKSPACE

echo "+ initialize npm in workspace"
npm init -y

echo "+ install test-target module in workspace"
npm i $PKG_PACKED

echo "+ run test-target module"
npx webmanifest --icon $EXAMPLE_ICON

echo "+ back to home"
cd ..


src/index.ts
#!/usr/bin/env node

import { pathExists, readJSON, writeJSON } from 'fs-extra'
import { imageSize } from 'image-size'
import meow from 'meow'
import sharp from 'sharp'

const ICON_SIZE_DELIMITER = ','
const MIN_ACCEPT_IMAGE_RESOLUTION = 512
const DEFAULT_INPUT_ICON_PATH = 'public/icon.svg'
const DEFAULT_INPUT_MANIFEST_PATH = 'public/manifest.webmanifest'
const DEFAULT_OUTPUT_ICON_PATH = 'public/assets'
const DEFAULT_OUTPUT_MANIFEST_PATH = 'public/manifest.webmanifest'
const DEFAULT_OUTPUT_SIZES = '512,384,192,180,152,144,128,96,72'

async function run() {
  const cli = meow(
    `
  Usage
    $ webmanifest --icon <filepath>
  Options
    --icon, -i Template icon file <required>
    --manifest, -m Template webmanifest file
    --outputIcon, -oi Output icons directory path
    --outputManifest, -om Output webmanifest directory path
    --sizes, -s Output icon sizes
  Examples
    $ webmanifest
    $ webmanifest --icon public/icon.svg --manifest public/manifest.webmanifest --outputIcon public/assets --outputManifest public/manifest.webmanifest --sizes 512,384,192,180,152,144,128,96,72
  `,
    {
      flags: {
        icon: {
          type: 'string',
          alias: 'i',
          default: DEFAULT_INPUT_ICON_PATH,
        },
        manifest: {
          type: 'string',
          alias: 'm',
          default: DEFAULT_INPUT_MANIFEST_PATH,
        },
        outputIcon: {
          type: 'string',
          alias: 'oi',
          default: DEFAULT_OUTPUT_ICON_PATH,
        },
        outputManifest: {
          type: 'string',
          alias: 'om',
          default: DEFAULT_OUTPUT_MANIFEST_PATH,
        },
        sizes: {
          type: 'string',
          alias: 's',
          default: DEFAULT_OUTPUT_SIZES,
        },
      },
    }
  )

  if (!(await pathExists(cli.flags.icon))) {
    throw new Error(`${cli.flags.icon} does not exists.`)
  }
  if (!(await pathExists(cli.flags.manifest))) {
    throw new Error(`${cli.flags.manifest} does not exists.`)
  }
  if (!(await pathExists(cli.flags.outputIcon))) {
    throw new Error(`${cli.flags.outputIcon} does not exists.`)
  }
  if (!(await pathExists(cli.flags.outputManifest))) {
    throw new Error(`${cli.flags.outputManifest} does not exists.`)
  }

  const iconRectangle = imageSize(cli.flags.icon)

  if (!iconRectangle.width || !iconRectangle.height) {
    throw new Error('Icon file is invalid.')
  }
  if (iconRectangle.width !== iconRectangle.height) {
    throw new Error('Icon file is required its square.')
  }
  if (
    iconRectangle.width &&
    iconRectangle.width < MIN_ACCEPT_IMAGE_RESOLUTION
  ) {
    throw new Error(
      `Icon file is required larger than ${MIN_ACCEPT_IMAGE_RESOLUTION}px.`
    )
  }
  if (
    iconRectangle.height &&
    iconRectangle.height < MIN_ACCEPT_IMAGE_RESOLUTION
  ) {
    throw new Error(
      `Icon file is required larger than ${MIN_ACCEPT_IMAGE_RESOLUTION}px.`
    )
  }

  const ascendingOrder = (a: number, b: number) => (a < b ? 1 : -1)

  const outputSizes = cli.flags.sizes.includes(ICON_SIZE_DELIMITER)
    ? cli.flags.sizes
        .split(ICON_SIZE_DELIMITER)
        .map(Number)
        .filter(Number)
        .sort(ascendingOrder)
    : [Number(cli.flags.sizes)].filter(Number)

  for (const size of outputSizes) {
    await sharp(cli.flags.icon)
      .resize(size, size)
      .png()
      .toFile(`${cli.flags.outputIcon}/icon-${size}x${size}.png`)
    console.log(`Output icon: ${cli.flags.outputIcon}/icon-${size}x${size}.png`)
  }

  const inputManifest = await readJSON(cli.flags.manifest)
  const outputManifestContent = {
    ...inputManifest,
    icons: outputSizes.map((size) => ({
      src: `icon-${size}x${size}.png`,
      sizes: `${size}x${size}`,
      type: 'image/png',
    })),
  }

  await writeJSON(cli.flags.outputManifest, outputManifestContent, {
    spaces: 2,
  })
  console.log(`Output webmanifest: ${cli.flags.outputManifest}`)

  console.log(
    '\n🎉 All icons and webmanifest file is generated successfully.\n'
  )
}

run()
